apps/communications/contacts/test/unit/views/details/details_controller_test.js

```diff
@@ -0,0 +1,135 @@
+  });
+
+  suiteTeardown(function() {
+    navigator.mozL10n = realMozL10n;
+    window.ContactToVcardBlob = realContactToVcardBlob;
+    realMozL10n = null;
+    realContactToVcardBlob = null;
+  });
+
+  setup(function() {
+    mozContact = new MockContactAllFields();
+    stubContactServiceSave = this.sinon.stub(ContactsService, 'save',
+                                              function(contact, callback) {});
+    stubLazyLoader = this.sinon.stub(LazyLoader, 'load',
+      function(files, callback) {
+      callback();
+    });
+  });
+
+  teardown(function() {
+    stubLazyLoader = null;
+    stubLazyLoader = null;
+    stubContactServiceSave = null;
+    mozContact = null;
+  });
+
+  suite('Close button', function() {
+    test(' > must call handleBackAction when an event is received',
+      function(done) {
+        var spy = this.sinon.spy(window.history, 'back');
+        window.addEventListener('backAction', function() {
+          assert.isTrue(spy.calledOnce);
+          done();
+        });
+
+        window.dispatchEvent(new CustomEvent('backAction'));
+    });
+  });
+
+  suite('Edit button', function() {
+    test('> must call handleEditAction when an event is received', function() {
+      //TODO: add tests when edit mode works
+      assert.isTrue(true);
+    });
+  });
+
+  suite('Share', function() {
+    var spyContactToVcardBlob;
+    setup(function() {
+      spyContactToVcardBlob = this.sinon.spy(window, 'ContactToVcardBlob');
+    });
+
+    teardown(function() {
+      spyContactToVcardBlob = null;
+    });
+
+    test(' > Must handle shareAction event', function(done) {
+      window.addEventListener('shareAction', function(evt) {
+        assert.equal(evt.detail.contact, mozContact);
+        assert.isTrue(spyContactToVcardBlob.calledOnce);
+        done();
+      });
+
+      window.dispatchEvent(new CustomEvent('shareAction', {'detail': {
+        'contact': mozContact
+      }}));
+    });
+  });
+
+  suite('Find Duplicates', function() {
+    test(' > Must handle find duplicates event', function(done) {
+      var stubMatch = this.sinon.stub(MatchService, 'match');
+      window.addEventListener('findDuplicatesAction', function() {
+        assert.isTrue(stubMatch.calledWith(mozContact.id));
+        done();
+      });
+
+      window.dispatchEvent(new CustomEvent('findDuplicatesAction', {'detail': 
+        {'contactId': mozContact.id}}));
+    });
+  });
+
+  suite('Toggle favorite', function() {
+    test(' > Must handle toggle favorite event', function(done) {
+      window.addEventListener('toggleFavoriteAction', function() {
+        assert.isTrue(stubContactServiceSave.calledOnce);
+        done();
+      });
+
+      function isFavorite()
+      {
+        return mozContact != null && mozContact.category != null &&
+              mozContact.category.indexOf('favorite') != -1;
+      }
+
+      window.dispatchEvent(new CustomEvent('toggleFavoriteAction', {'detail': 
```

Comment:
We cannot test if the |toggleFavorite| method of our controller is called because it is a private function. However, we can check if ContactsServices.save is called, and that is what we test in this suit.